### PR TITLE
Improve client secret update logic

### PR DESCRIPTION
**Pull request by thiva-k**
This pull request improves the handling of OAuth client secrets during application updates, ensuring that secrets are preserved or updated correctly based on the input and existing application state. It introduces a new helper function to encapsulate this logic and adds comprehensive unit tests to cover various scenarios.

**OAuth client secret handling improvements:**

* Added a new function `getProcessedClientSecretForUpdate` to determine the correct hashed client secret during update operations, preserving the existing secret if no new secret is provided and handling public clients appropriately.
* Updated the `UpdateApplication` service method to use `getProcessedClientSecretForUpdate` instead of the previous helper, ensuring correct behavior during updates.
* Enhanced the logic in `processInboundAuthConfig` to avoid regenerating a client secret if the existing confidential OAuth app already has a secret, further preserving secrets during updates.

**Testing enhancements:**

* Added multiple unit tests for `getProcessedClientSecretForUpdate`, covering cases such as public clients, new secrets provided, preservation of existing secrets, and scenarios with no existing app or OAuth config.

## Summary by CodeRabbit

* **Bug Fixes**
  * Prevented accidental resetting of OAuth/OIDC application credentials during updates; existing confidential client secrets are preserved when no new secret is provided, and new secrets are generated only when required.

* **Tests**
  * Added unit tests covering public vs. confidential clients, new-secret handling, preservation of existing hashed secrets, and various edge cases with missing or changing configuration.

Related Issue:

- https://github.com/asgardeo/thunder/issues/1399

### CODE REVIEW

File: backend/internal/application/service.go [hunk 1]
```diff
@@ -930,12 +931,21 @@ func (as *applicationService) processInboundAuthConfig(app *model.ApplicationDTO
...
-		generatedClientSecret, err := oauthutils.GenerateOAuth2ClientSecret()
-		if err != nil {
-			logger.Error("Failed to generate OAuth client secret", log.Error(err))
-			return nil, &ErrorInternalServerError
+		// Check if we should preserve existing confidential OAuth config secret
+		shouldPreserveSecret := existingApp != nil &&
+			len(existingApp.InboundAuthConfig) > 0 &&
+			existingApp.InboundAuthConfig[0].OAuthAppConfig != nil &&
+			existingApp.InboundAuthConfig[0].OAuthAppConfig.HashedClientSecret != "" &&
+			!existingApp.InboundAuthConfig[0].OAuthAppConfig.PublicClient
```
Comment (rajithacharith) [inline backend/internal/application/service.go]: Calling the 0th element is wrong design we have followed. 
InboundAuthConfig is an array as we are planning to support other inbound protocols as SAML. 
Then there should be a fn to getOAuthConfigs something. 
Otherwise this pattern will be a trouble when we are onboarding SAML
This might not need fix in this PR, but shall we keep track?

File: backend/internal/application/service.go [hunk 2]
```diff
@@ -1518,3 +1528,30 @@ func getProcessedClientSecret(oauthConfig *model.OAuthAppConfigDTO) string {
 	}
 	return hash.GenerateThumbprintFromString(oauthConfig.ClientSecret)
 }
+
+// getProcessedClientSecretForUpdate returns the hashed client secret for update operations.
+// If a new secret is provided, it hashes it. Otherwise, it preserves the existing hashed secret.
+func getProcessedClientSecretForUpdate(
+	oauthConfig *model.OAuthAppConfigDTO,
+	existingApp *model.ApplicationProcessedDTO,
```
Comment (rajithacharith) [inline backend/internal/application/service.go]: Do we really need this existingApp variable?

File: backend/internal/application/service.go (hunk 1 above)
Comment (thiva-k) [reply backend/internal/application/service.go]: Ack. Will be tracked and updated with the planned application improvements

File: backend/internal/application/service.go (hunk 2 above)
Comment (thiva-k) [reply backend/internal/application/service.go]: Yes. To get the existing hashed client secret for updating apps that do not have client secret in payload.

File: backend/internal/application/service.go [hunk 3]
```diff
@@ -930,12 +931,21 @@ func (as *applicationService) processInboundAuthConfig(app *model.ApplicationDTO
 	// Generate OAuth 2.0 compliant client secret with high entropy for security
 	// Only generate client secret for confidential clients
 	if inboundAuthConfig.OAuthAppConfig.ClientSecret == "" && !inboundAuthConfig.OAuthAppConfig.PublicClient {
-		generatedClientSecret, err := oauthutils.GenerateOAuth2ClientSecret()
-		if err != nil {
-			logger.Error("Failed to generate OAuth client secret", log.Error(err))
-			return nil, &ErrorInternalServerError
+		// Check if we should preserve existing confidential OAuth config secret
+		shouldPreserveSecret := existingApp != nil &&
```
Comment (ThaminduDilshan) [inline backend/internal/application/service.go]: Shall we extract out this client secret resolution logic into a separate private method to improve code readability?

File: backend/internal/application/service.go (hunk 2 above)
Comment (thiva-k) [reply backend/internal/application/service.go]: Improved to pass only the existing app's oauth client as parameter

File: backend/internal/application/service.go (hunk 3 above)
Comment (thiva-k) [reply backend/internal/application/service.go]: Updated

File: backend/internal/application/service.go [hunk 4]
```diff
@@ -1518,3 +1518,62 @@ func getProcessedClientSecret(oauthConfig *model.OAuthAppConfigDTO) string {
...
+	return ""
+}
+
+// resolveClientSecret generates a new client secret for confidential clients if needed.
+// It preserves existing secrets during update operations unless explicitly provided.
+func resolveClientSecret(
+	inboundAuthConfig *model.InboundAuthConfigDTO,
+	existingApp *model.ApplicationProcessedDTO,
+) *serviceerror.ServiceError {
+	logger := log.GetLogger().With(log.String(log.LoggerKeyComponentName, "ApplicationService"))
```
Comment (ThaminduDilshan) [inline backend/internal/application/service.go:1552]: Shall we use the application service logger instead?

Comment (ThaminduDilshan) [reply backend/internal/application/service.go:1552]: You'll have to convert this into a private service method

Comment (thiva-k) [reply backend/internal/application/service.go:1552]: Application do not have a service level logger as of now. Will be added in future improvements
